### PR TITLE
Port cuDNN RNN dropout state initialization to ATen and make Python c…

### LINKED ISSUE #5138: Bind cuDNN dropout initialization from ATen
In `torch/backends/cudnn/rnn.py` we have the last vestiges of the Python-side cuDNN bindings:

```
class Unserializable(object):

    def __init__(self, inner):
        self.inner = inner

    def get(self):
        return self.inner

    def __getstate__(self):
        # Note: can't return {}, because python2 won't call __setstate__
        # if the value evaluates to False
        return "<unserializable>"

    def __setstate__(self, state):
        self.inner = None


def init_dropout_descriptor(handle, dropout, train, dropout_seed, dropout_state):
    dropout_desc_name = 'desc_' + str(torch.cuda.current_device())
    dropout_p = dropout if train else 0
    if (dropout_desc_name not in dropout_state) or (dropout_state[dropout_desc_name].get() is None):
        dropout_state[dropout_desc_name] = Unserializable(
            cudnn.DropoutDescriptor(handle, dropout_p, dropout_seed)
        )
    dropout_desc = dropout_state[dropout_desc_name].get()
    dropout_desc.set_dropout(dropout_p, dropout_seed)
    return dropout_desc


def get_dropout_state(fn, handle):
    dropout_desc_name = 'desc_' + str(torch.cuda.current_device())
    dropout_p = fn.dropout if fn.train else 0
    dropout_desc = fn.dropout_state[dropout_desc_name].get()
    return dropout_desc.state
```

We'd like to get rid of this code entirely, and that means we need to allocate the tensor state in `dropout_desc` from ATen.

The code to do this already exists in `aten/src/ATen/cudnn/Descriptors.h`; it's just a matter of taking dropout parameters and returning an initialized dropout state buffer, which can be passed to subsequent ATen calls.

### PR DESCRIPTION
**Pull request by ezyang**
…ode use it.

Fixes #5138.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

CC @ebetica; this changes the API so instead of calling the initialization function with `input` you just call it with `input.type()`.

### CODE REVIEW

File: torch/backends/cudnn/rnn.py [hunk 1]
```diff
@@ -35,20 +35,13 @@ def __setstate__(self, state):
...
-            cudnn.DropoutDescriptor(handle, dropout_p, dropout_seed)
+            torch._C._VariableFunctions._cudnn_init_dropout_state(ty, dropout_p, train, dropout_seed)
+            if dropout_p != 0 else None
         )
-    dropout_desc = dropout_state[dropout_desc_name].get()
-    dropout_desc.set_dropout(dropout_p, dropout_seed)
-    return dropout_desc
-
-
-def get_dropout_state(fn, handle):
```
Comment (gchanan) [inline torch/backends/cudnn/rnn.py]: we don't need this for backwards compatibility too?

File: torch/backends/cudnn/rnn.py [hunk 2]
```diff
@@ -35,20 +35,13 @@ def __setstate__(self, state):
...
 
 
-def init_dropout_descriptor(handle, dropout, train, dropout_seed, dropout_state):
+def init_dropout_state(ty, dropout, train, dropout_seed, dropout_state):
     dropout_desc_name = 'desc_' + str(torch.cuda.current_device())
     dropout_p = dropout if train else 0
     if (dropout_desc_name not in dropout_state) or (dropout_state[dropout_desc_name].get() is None):
         dropout_state[dropout_desc_name] = Unserializable(
-            cudnn.DropoutDescriptor(handle, dropout_p, dropout_seed)
+            torch._C._VariableFunctions._cudnn_init_dropout_state(ty, dropout_p, train, dropout_seed)
```
Comment (gchanan) [inline torch/backends/cudnn/rnn.py]: fwiw you can probably do torch._cudnn_init_dropout.

**Review by gchanan — APPROVED** [review]
I think is fine; I'm changing how these functions get bound (mainly so they get an at:: namespace function and the Type that is passed from Type is *this rather than using a separate parameter), but I don't think this should break anything.

File: torch/backends/cudnn/rnn.py (hunk 1 above)
Comment (ezyang) [reply torch/backends/cudnn/rnn.py]: I was under the impression that `torch.backends.cudnn.rnn` was private. If it wasn't, I've got a lot more things I've got to bring back...

File: torch/backends/cudnn/rnn.py (hunk 2 above)
Comment (ezyang) [reply torch/backends/cudnn/rnn.py]: OK, I'll fix this and a bunch of others in its own commit.
